kedro/io/kedro_data_catalog.py

```diff
@@ -112,6 +114,50 @@ def __eq__(self, other) -> bool:  # type: ignore[no-untyped-def]
             other.config_resolver.list_patterns(),
         )
 
+    def keys(self, regex_search: str | None = None) -> List[str]:  # noqa: UP006
+        return self._filter_keys(regex_search)
+
+    def values(self, regex_search: str | None = None) -> List[AbstractDataset]:  # noqa: UP006
+        return [self._datasets[key] for key in self._filter_keys(regex_search)]
+
+    def items(
+        self, regex_search: str | None = None
+    ) -> List[tuple[str, AbstractDataset]]:  # noqa: UP006
+        return [(key, self._datasets[key]) for key in self._filter_keys(regex_search)]
+
+    def __iter__(self) -> str:
+        yield from self._datasets.keys()
+
+    def __getitem__(self, ds_name: str) -> AbstractDataset | None:
+        return self.get_dataset(ds_name)
+
+    def __setitem__(self, key: str, value: Any) -> None:
+        if key in self._datasets:
+            self._logger.warning("Replacing dataset '%s'", key)
+        if isinstance(value, AbstractDataset):
+            self._datasets[key] = value
+        else:
+            self._datasets[key] = MemoryDataset(data=value)  # type: ignore[abstract]
```

Comment:
I really (really) think we should have at least logger.info() here to explain this behaviour. I expect silent error due to people assigning a malformed dataset manually which end up assigned as a MemoryDataset, opening a room for a lot of headache.

Reply:
I don't mind adding it, though previously, we didn't log it in `add_feed_dict()`, and there were no common issues with that. However, it could be because not many people used this method.

Reply:
Wow, I did not know it was already possible ; however as you say I think we target notebooks users with this design, so we can expect it to happen more often that it does now.